### PR TITLE
Add aten export for embedding_bag

### PR DESCRIPTION
**Pull request by ashbhandare**
Embedding_bag export for inference contains Loop, which we don't have a gradient for. This PR exports embedding_bag as Aten op and adds gradient.

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/cpu/aten_op_executor/aten_op_executor.cc [hunk 1]
```diff
@@ -70,7 +95,20 @@ struct ATenOperator {
...
+      // sometimes the tensor does not contain any data. eg: the maxindices output 
+      // for embedding_bag_backward cuda kernel. In such cases, fromDLPack() errors out
+      // while trying to make a tensor while trying to match the data's device with 
+      // the device specified in dlpack.
+      // so skipping using fromDLPack in such cases to create empty tensor.
+      at::Device device = getATenDevice(dlpack->dl_tensor.device);
+      at::ScalarType stype = at::toScalarType(dlpack->dl_tensor.dtype);
+      tensor = at::empty(at::IntArrayRef(dlpack->dl_tensor.shape, dlpack->dl_tensor.ndim), 
+                         at::device(device).dtype(stype));
+    }
```
Comment (iK1D) [inline orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/cpu/aten_op_executor/aten_op_executor.cc:111]: You may need to call "dlpack->deleter(const_cast<DLManagedTensor*>(dlpack));" here, otherwise the dlpack will leak.

**Review by iK1D — CHANGES_REQUESTED** [review]

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/cpu/aten_op_executor/aten_op_executor.cc [hunk 2]
```diff
@@ -44,6 +44,31 @@ c10::IValue Int64ToBoolIValue(const DLManagedTensor* dlpack, bool is_list, bool
   return is_optional ? c10::IValue(c10::optional<bool>(value)) : c10::IValue(value);
 }
 
+// function from pytorch/aten/src/ATen/DLConvertor.cpp
+static at::Device getATenDevice(const DLDevice& ctx) {
+  switch (ctx.device_type) {
+    case DLDeviceType::kDLCPU:
+      return at::Device(at::DeviceType::CPU);
+#ifndef USE_ROCM
```
Comment (iK1D) [inline orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/cpu/aten_op_executor/aten_op_executor.cc:52]: I saw the build pipelien is failed for AMD. In this file, we don't have USE_ROCM macro defined, so using marco here is not correct. I feel like we should not copy code from PyTorch to here, it's hard for us to maintain in the future, and the device logic is not same compared to ORT. In ORT, we have only GPU device, if USE_ROCM then dlpack is kDLROCM, otherwise it's kDLCUDA. But in PyTorch, from the logic, if USE_ROCM is defined, kDLROCM will also map to DeviceType::CUDA, otherwise it's DeviceType::HIP. I am not quite understand the logic here, but I feel like if it's from ORT->DLPack->Torch, I think we can simply map both kDLCUDA and kDLROCM to Torch's DeviceType::CUDA, there is no mapping from ORT's any device to Torch's DeviceType::HIP.

Comment (ashbhandare) [reply orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/cpu/aten_op_executor/aten_op_executor.cc]: changed

**Review by iK1D — APPROVED** [review]
Looks good. Need to format the code to align with the Lint requirement added recently.

**Comment by justinchuby** [comment]
Could you fix the python lint errors? Thanks!
